x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/geo_threshold.ts

```diff
@@ -0,0 +1,270 @@
+          const docId = _.get(subBucket, `entityHits.hits.hits[0]._id`);
+
+          return {
+            location,
+            shapeLocationId: bucketKey,
+            entityName: subBucket.key,
+            dateInShape,
+            docId,
+          };
+        });
+      })
+      .orderBy(['entityName', 'dateInShape'], ['asc', 'desc'])
+      .sortedUniqBy('entityName')
+      .value()
+  );
+}
+
+interface EntityMovementDescriptor {
+  entityName: string;
+  currLocation: {
+    location: number[];
+    shapeId: string;
+    date: string | null;
+    docId: string;
+  };
+  prevLocation: {
+    location: number[];
+    shapeId: string;
+    date: string | null;
+    docId: string;
+  };
+}
+
+export function getMovedEntities(
+  currLocationArr: LatestEntityLocation[],
+  prevLocationArr: LatestEntityLocation[],
+  trackingEvent: string
+): EntityMovementDescriptor[] {
+  return (
+    currLocationArr
+      // Check if shape has a previous location and has moved
+      .reduce(
+        (
+          accu: EntityMovementDescriptor[],
+          {
+            entityName,
+            shapeLocationId,
+            dateInShape,
+            location,
+            docId,
+          }: {
+            entityName: string;
+            shapeLocationId: string;
+            dateInShape: string | null;
+            location: number[];
+            docId: string;
+          }
+        ) => {
+          const prevLocationObj = prevLocationArr.find(
+            (locationObj: LatestEntityLocation) => locationObj.entityName === entityName
+          );
+          if (!prevLocationObj) {
+            return accu;
+          }
+          if (shapeLocationId !== prevLocationObj.shapeLocationId) {
+            accu.push({
+              entityName,
+              currLocation: {
+                location,
+                shapeId: shapeLocationId,
+                date: dateInShape,
+                docId,
+              },
+              prevLocation: {
+                location: prevLocationObj.location,
+                shapeId: prevLocationObj.shapeLocationId,
+                date: prevLocationObj.dateInShape,
+                docId: prevLocationObj.docId,
+              },
+            });
+          }
+          return accu;
+        },
+        []
+      )
+      // Do not track entries to or exits from 'other'
+      .filter((entityMovementDescriptor: EntityMovementDescriptor) =>
+        trackingEvent === 'entered'
+          ? entityMovementDescriptor.currLocation.shapeId !== OTHER_CATEGORY
+          : entityMovementDescriptor.prevLocation.shapeId !== OTHER_CATEGORY
+      )
+  );
+}
+
+export const getGeoThresholdExecutor = ({ logger: log }: { logger: Logger }) =>
+  async function ({
+    previousStartedAt: currIntervalStartTime,
+    startedAt: currIntervalEndTime,
+    services,
+    params,
+    alertId,
+    state,
+  }: {
+    previousStartedAt: Date | null;
+    startedAt: Date;
+    services: AlertServices;
+    params: {
```

Comment:
It would be better to use the config-schema generated type here, rather than copy it by hand, since if you change the schema, it will automatically update the generated TS type.  There's an example here: https://github.com/elastic/kibana/blob/63ff0606df1d35978049ff9aa25f04c46be895f8/x-pack/plugins/alerting_builtins/server/alert_types/index_threshold/alert_type_params.ts#L14-L28

The `Params` typescript type matches what the schema validates, and an equivalent type for your params can completely replace the list of properties in this `params` object definition.